rust/arrow/src/array/array.rs

```diff
@@ -1598,6 +1601,229 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
     }
 }
 
+/// A dictonary array where each element is a single value indexed by an integer key.
+pub struct DictionaryArray {
+    data: ArrayDataRef,
+    values: ArrayRef,
+    raw_values: RawPtrBox<u8>,
+    key_type: DataType,
+}
+
+macro_rules! collect_values {
```

Comment:
I think we should be able to use the `take` compute kernel here. `take` takes an array and an index of locations to index from. Such that you loosely get the below pseudo:

```rust
let array = ["a", "c", "b"];
let indices = [0, 1, 0, null, 2, 1]
let result = take(aray, indices)
// result would be ["a", "c", "a", null, "b", "c"]
```

If the take's index type isn't what you have, you can use the `cast` kernel to convert the type to what `take` accepts.

Reply:
Good idea. I'll read more about the compute kernels.